src/main/java/org/apache/mesos/scheduler/MesosToSchedulerDriverAdapter.java

```diff
@@ -0,0 +1,731 @@
+package org.apache.mesos.scheduler;
+
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.protobuf.ByteString;
+import org.apache.mesos.protobuf.Devolver;
+import org.apache.mesos.protobuf.Evolver;
+import org.apache.mesos.v1.scheduler.V1Mesos;
+import org.apache.mesos.v1.scheduler.Mesos;
+import org.apache.mesos.v1.scheduler.Protos;
+import org.apache.mesos.v1.scheduler.V0Mesos;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+
+/**
+ * This is a threadsafe  adapter from the new v1 `Mesos` + `Scheduler` interface to the old  v0 `SchedulerDriver`
+ * + `Scheduler` interface. It intercepts:
+ * - The v1 scheduler callbacks and converts them into appropriate v0 scheduler callbacks.
+ * - The various `driver.xx()` calls, creates a `Call` message and then invokes `send()` on the v1
+ * `Mesos` interface.
+ */
+public class MesosToSchedulerDriverAdapter implements
+        org.apache.mesos.v1.scheduler.Scheduler,
+        org.apache.mesos.SchedulerDriver {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MesosToSchedulerDriverAdapter.class);
+
+    // State of the `Mesos` interface.
+    private enum State {
+        DISCONNECTED,
+        CONNECTED,
+        SUBSCRIBED
+    }
+
+    // Exponential back-off between consecutive subscribe calls.
+    private static final int MULTIPLIER = 2;
+    private static final int SEED_BACKOFF_MS = 2000;
+    private static final int MAX_BACKOFF_MS = 30000;
+
+    private org.apache.mesos.Scheduler wrappedScheduler;
+    private org.apache.mesos.v1.Protos.FrameworkInfo frameworkInfo;
+    private final String master;
+    private final org.apache.mesos.v1.Protos.Credential credential;
+    // TODO(anand): This can change to `v1.Status` once we add support for devolving enums.
+    private org.apache.mesos.Protos.Status status;
```

Comment:
@gabrielhartmann `status` is managed internally, and is not supposed to be set from an outside entity. 

`status` can only be modified by the constructor(s), start, stop and abort methods. And, they are all synchronized, so we'll not run into a concurrent modification scenario.